core/src/main/java/org/elasticsearch/cluster/routing/OperationRouting.java

```diff
@@ -177,10 +178,17 @@ private ShardIterator preferenceActiveShardIterator(IndexShardRoutingTable index
             }
         }
         // if not, then use it as the index
+        int routingHash = Murmur3HashFunction.hash(preference);
+        if (nodes.getMinNodeVersion().onOrAfter(Version.V_6_0_0_alpha1_UNRELEASED)) {
+            // Better overall balancing can be achieved if each shardId opts
+            // for a different selection from the elements in the list
+            // given the same user-supplied preference key.
+            routingHash = 31 * routingHash + indexShard.shardId.id();
```

Comment:
can we use `indexShard.shardId.hashCode()` - it's good to have different hashing across indices as well.